### PR TITLE
Improve queue

### PR DESCRIPTION
**Pull request by lstonussi**
### Summary by CodeRabbit

- New Feature: Introduced a more intuitive way to manage media queues with the addition of `removeByPosition` method, allowing users to remove specific items from the queue.
- Improvement: Enhanced the media queue's handling of current and next items, providing a smoother user experience when navigating through the queue.
- Refactor: Streamlined the internal workings of the media queue, improving the reliability and performance of the media player.
- Test: Updated test cases to reflect the new functionalities and improvements, ensuring the robustness of the application.
- Bug Fix: Fixed potential errors when reaching the end of the queue, enhancing the stability of the media player.

### CODE REVIEW

File: packages/player/lib/src/queue.dart [hunk 1]
```diff
@@ -153,45 +168,53 @@ class Queue {
 
   remove(Media media) {
     final itemToBeRemoved = storage.firstWhere((i) => i.item.id == media.id);
+    storage.remove(itemToBeRemoved);
     if (itemToBeRemoved.position < index) {
-      --index;
+      setIndex = index - 1;
     }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:174]: lembra que falamos de fazer por indice?

File: packages/player/lib/src/queue.dart [hunk 2]
```diff
@@ -153,45 +168,53 @@ class Queue {
...
-    storage.remove(itemToBeRemoved);
     for (var i = itemToBeRemoved.position; i < storage.length; ++i) {
-      storage[i].position -= 1;
-      storage[i].originalPosition -= 1;
+      storage[i].position--;
+      if (storage[i].originalPosition > 0) {
+        storage[i].originalPosition--;
+      }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:179]: tem que verificar toda a fila e a original position do que esta removendo. isso aqui vai continuar com erro

File: packages/player/lib/src/queue.dart [hunk 3]
```diff
@@ -35,7 +35,19 @@ class Queue {
...
+  int get index => _index;
+  Media? _current;
+
+  set setIndex(int index) {
+    _index = index;
+    if (storage.length > 0 && index >= 0) {
+      _current = storage[index].item;
+    } else {
+      return null;
+    }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:48]: set nao tem retorno

File: packages/player/lib/src/queue.dart [hunk 4]
```diff
@@ -318,38 +340,50 @@ class Queue {
...
     final playingItem = storage.elementAt(index);
-
     if (newIndex > oldIndex) {
-      for (int i = oldIndex + 1; i <= newIndex; i++) {
-        if (!isShuffle) {
-          storage.elementAt(i).originalPosition--;
+      for (int i = oldIndex; i <= newIndex; i++) {
+        if (!isShuffle && storage[i].originalPosition > 0) {
+          storage[i].originalPosition--;
         }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:353]: igual acima, tem que ser um for inteiro quando ta em shuffle , igual nao faz sentido esse isShuflle pois sempre tem que acertar o origianl, igual abaixo tb

**Review by atrope — CHANGES_REQUESTED** [review]

File: packages/player/lib/src/queue.dart (hunk 1 above)
Comment (lstonussi) [reply packages/player/lib/src/queue.dart:174]: Por indice nao rolou pq iriamos ter que fazer a mesma coisa `    final itemToBeRemoved = storage.firstWhere((i) => i.item.id == media.id);
`
pq o position fica atrasado no loop de delecao

File: packages/player/lib/src/queue.dart [hunk 5]
```diff
@@ -151,47 +164,66 @@ class Queue {
...
-    storage.remove(itemToBeRemoved);
-    for (var i = itemToBeRemoved.position; i < storage.length; ++i) {
-      storage[i].position -= 1;
-      storage[i].originalPosition -= 1;
+    if (!isShuffle) {
+      for (var i = itemToBeRemoved.position; i < storage.length; ++i) {
+        storage[i].position--;
+        if (storage[i].originalPosition > 0) {
+          storage[i].originalPosition--;
+        }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:178]: se nao for shuffle, nao precisa desse if. 


se estiver removendo index 1 para cima, ele nao sera 0 .
Se esrtiver removendo o 0, todos acima dele nao serao 0.

File: packages/player/lib/src/queue.dart [hunk 6]
```diff
@@ -151,47 +164,66 @@ class Queue {
...
+      }
+    } else {
+      for (var i = 0; i < storage.length; ++i) {
+        storage[i].position = i;
+        if (storage[i].originalPosition > 0 &&
+            storage[i].originalPosition > itemToBeRemoved.originalPosition) {
+          storage[i].originalPosition--;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (atrope) [reply packages/player/lib/src/queue.dart:188]: bom ter essee if tb

File: packages/player/lib/src/queue.dart [hunk 7]
```diff
@@ -202,7 +234,7 @@ class Queue {
 
   Media rewind() {
     assert(index >= 0);
-    return storage.elementAt(index).item;
+    return storage[index].item;
   }
```
[the remark this answers is not in the document]
Comment (atrope) [reply packages/player/lib/src/queue.dart:238]: boa

File: packages/player/lib/src/queue.dart [hunk 8]
```diff
@@ -35,7 +35,17 @@ class Queue {
...
+  var _index = 0;
+  int get index => _index;
+  Media? _current;
+
+  set setIndex(int index) {
+    if (storage.isNotEmpty && index >= 0) {
+      _index = index;
+      _current = storage[index].item;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (atrope) [reply packages/player/lib/src/queue.dart:47]: boa

**Review by atrope — CHANGES_REQUESTED** [review]

File: packages/player/lib/src/queue.dart [hunk 9]
```diff
@@ -151,47 +167,105 @@ class Queue {
...
+        for (var e in storage) {
+          debugPrint(
+              '=====> storage remove: ${e.item.name} - ${e.position} | ${e.originalPosition}');
+        }
+      }
+    } catch (e) {
+      return 0;
     }
+    return 1;
   }
```
Comment (atrope) [inline packages/player/lib/src/queue.dart:232]: acha a posicao e usa o RemoveByPosition,

**Review by atrope — CHANGES_REQUESTED** [review]

**Review by atrope — APPROVED** [review]

**Review by atrope — APPROVED** [review]

**Review by felipecastrosales — APPROVED** [review]

**Pull request merged**